### PR TITLE
Move release under test-deploy workflow

### PR DESCRIPTION
**Pull request by fotos**
Move the `release` step under the `test-deploy` workflow.

Previously the (setup) workflow [failed with](https://app.circleci.com/pipelines/gh/react-native-community/react-native-circleci-orb/359):

> Max number of workflows exceeded.

Turns out [dynamic configurations](https://circleci.com/docs/dynamic-config/) that use `setup: true` can have only [one workflow defined](https://support.circleci.com/hc/en-us/articles/360060934851-Troubleshooting-Error-Message-Max-number-of-workflows-exceeded).

Now the idea for the flow is:

```mermaid
graph LR;
    merge["Merge to master"]
    lint["lint job"]
    pack["pack job"]
    review["review job"]
    check["check job"]
    continue["continue job"]
    command-test["command-test job"]
    semantic-release["semantic-release job"]
    tag["Tag release"]

    merge --> CI
    CI --> tag

    subgraph CI
      direction LR
      subgraph lint-pack
        lint --> continue
        pack --> continue
        review --> continue
        check --> continue
      end

      subgraph test-deploy
        continue --> command-test
        command-test --> semantic-release
      end
    end
```

After the tag is pushed to `GitHub` this triggers another CI run with:

```mermaid
graph LR;
    tag["Tag release"]
    lint["lint job"]
    pack["pack job"]
    review["review job"]
    check["check job"]
    continue["continue job"]
    command-test["command-test job"]
    pack["pack job"]
    pack-release["pack job"]
    publish["publish job"]
    orb["Publish Orb"]

    tag --> CI
    CI --> orb

    subgraph CI
      direction LR
      subgraph lint-pack
        lint --> continue
        pack --> continue
        review --> continue
        check --> continue
      end

      subgraph test-deploy
        continue --> command-test
        continue --> pack-release
        command-test --> publish
        pack-release --> publish
      end
    end
```

_(My 🧜‍♀️ diagramming sucks but hopefully you get the idea)_

This PR is a follow up to #177.

### CODE REVIEW

**Comment by fotos** [comment]
@cortinico look into the [current releases](https://github.com/react-native-community/react-native-circleci-orb/releases) I noticed they have been created by the [`react-native-community-bot`](https://github.com/react-native-community-bot). My gut feeling tells me that I shouldn't have removed the `GH_TOKEN` from the ENV because that's most likely belonged to `react-native-community-bot` and was used to push the releases back to GitHub.

Do you have access to the `react-native-community-bot` account? We might need a new token for `semantic-release`.

**Comment by fotos** [comment]
> @cortinico looking into the [current releases](https://github.com/react-native-community/react-native-circleci-orb/releases) I noticed they have been created by the [`react-native-community-bot`](https://github.com/react-native-community-bot). My gut feeling tells me that I shouldn't have removed the `GH_TOKEN` from the ENV because that's most likely belonged to `react-native-community-bot` and was used to push the releases back to GitHub.
> 
> Do you have access to the `react-native-community-bot` account? We might need a new token for `semantic-release`.

Yes, looks like the `GH_TOKEN` is going to be [required by `semantic-release`](https://semantic-release.gitbook.io/semantic-release/usage/ci-configuration#authentication) in order to push the tags back. This is also [mentioned in the CircleCI documentation](https://circleci.com/docs/github-integration/#deploy-keys-and-user-keys).

> When you add a new project, CircleCI creates a deployment key on GitHub for your project. A deploy key is an SSH key-pair, one public, one private. GitHub stores the public key, and CircleCI stores the private key. The deployment key gives CircleCI access to a single repository. **To prevent CircleCI from pushing to your repository, this deployment key is read-only.**
> 
> If you want to push to the repository from your builds, you will **need a deployment key with write access**. See the below section for GitHub-specific instructions to create a deployment key.

_(emphasis mine)_

Currently the `react-native-community-bot` is acting as a machine.

I think we need to get access to the `react-native-community-bot` account, create a new PAT scoped with appropriate permissions for this project and then add the token as the `GH_TOKEN` environment variable. That should [allow `semantic-release` to push tags](https://semantic-release.gitbook.io/semantic-release/recipes/ci-configurations/circleci-workflows).

**Comment by cortinico** [comment]
> Currently the `react-native-community-bot` is acting as a machine.

Nope I don't have it.

How about we just don't use semantic-release for now to unblock the situation and then get back to setting it up in the future?

**Comment by fotos** [comment]
> > Currently the `react-native-community-bot` is acting as a machine.
> 
> Nope I don't have it.
> 
> How about we just don't use semantic-release for now to unblock the situation and then get back to setting it up in the future?

Sure. We can merge this PR and tag a release manually via GitHub UI.

**Review by cortinico — APPROVED** [review]

**Comment by fotos** [comment]
@cortinico this went better except the the `node/run` job which [failed as expected](https://app.circleci.com/pipelines/github/react-native-community/react-native-circleci-orb/361).

Will you create a new "tag" in GH UI or shall I do that?

**Comment by cortinico** [comment]
> Will you create a new tag in GH UI or shall I do that?

Please go ahead 🙏

**Comment by fotos** [comment]
> > Will you create a new tag in GH UI or shall I do that?
> 
> Please go ahead 🙏

Just pushed the `v7.4.0` [tag](https://github.com/react-native-community/react-native-circleci-orb/releases/tag/v7.4.0). [CI is running](https://app.circleci.com/pipelines/github/react-native-community/react-native-circleci-orb/363/workflows/785f0779-d6f2-492c-a76a-338976aa71e4).

Truth be told I'd rather have figured out the `GH_TOKEN` for `semantic-release`. What we did is an ugly workaround and there is quite some clean up to sort things out (CHANGELOG, versioning, GH releases etc.).

How can we find the credentials for `react-native-community-bot`?

Can I also get admin access to the project to sort out the CHANGELOG etc by running `semantic-release` locally and (re)doing the GH releases/tags?

**Comment by fotos** [comment]
@cortinico everything went well except [publishing didn't work](https://app.circleci.com/pipelines/github/react-native-community/react-native-circleci-orb/363/workflows/51463e4b-3d80-409b-8fdb-87b2bb79b672/jobs/577) because the `CIRLCE_TOKEN` environment variable is missing. To be honest at this point I'd rather get admin access to the project and do things the right way (™️) than continue to blindly hack-around. What do you think?

**Comment by fotos** [comment]
My understanding of how this publishing is supposed to work:

1. `orb-tools/publish` jobs [requires a `CIRCLE_TOKEN` environment variable](https://github.com/CircleCI-Public/orb-tools-orb/blob/v12.0.4/src/jobs/publish.yml#L32-L38)
2. This environment variable is supposed to be coming from the [`orb-publishing` context](https://github.com/react-native-community/react-native-circleci-orb/blob/05bd36f197e6930a70df1a4a9cd6ed5e521dc6a6/.circleci/test-deploy.yml#L53) ([template](https://github.com/CircleCI-Public/Orb-Template/blob/3b92805247791f9f5263c1eb8b09039eaac730cd/.circleci/test-deploy.yml#L50))

The documentation on `orb-tools/publish` mentions:

>    Enter the name of the environment variable containing your CircleCI API Token.
>    This token must have production publishing scope (organization owner).
>    Never directly enter the value of any secret tokens in your config.

So, it looks like we need secrets from the organization owner anyway.

**Comment by cortinico** [comment]
> Truth be told I'd rather have figured out the `GH_TOKEN` for `semantic-release`. What we did is an ugly workaround and there is quite some clean up to sort things out (CHANGELOG, versioning, GH releases etc.).

Can't we just use the GitHub Actions Token to run semantic-release?

> Can I also get admin access to the project to sort out the CHANGELOG etc by running semantic-release locally and (re)doing the GH releases/tags?

You should have access now 👍

**Comment by fotos** [comment]
> > Truth be told I'd rather have figured out the `GH_TOKEN` for `semantic-release`. What we did is an ugly workaround and there is quite some clean up to sort things out (CHANGELOG, versioning, GH releases etc.).
> 
> Can't we just use the GitHub Actions Token to run semantic-release?

I'll look into that now.

> > Can I also get admin access to the project to sort out the CHANGELOG etc by running semantic-release locally and (re)doing the GH releases/tags?
> 
> You should have access now 👍

Thank you – let me see what I can do.

By the way I'll gladly relinquish admin access as soon as we get the CI/CD working again.

**Comment by fotos** [comment]
@cortinico I ~~spent~~ invested 2 hours and this is where we stand:

## Publishing Orbs

I went through all the scripts in `orb-tools` to understand how an [orb version is published](https://github.com/CircleCI-Public/orb-tools-orb/blob/v12.0.4/src/scripts/publish.sh#L29). It boils down to two `circleci` commands. For example to publish version `7.2.0`:

```
$ git checkout v7.2.0
$ circleci orb pack src > dist/orb.yml
$ circleci orb publish dist/orb.yml react-native-community/react-native@7.2.0
```

To do the above requires a `CIRCLE_TOKEN`. What is the `CIRCLE_TOKEN` and how is it created? When following the [guide](https://circleci.com/docs/create-an-orb/#4-follow-the-prompts-to-set-up-your-orb) at the `circleci orb init` step, the setup process generates a CircleCI personal token and stores it in the `orb-publishing` context as the `CIRCLE_TOKEN` environment variable. This is described clearly in the guide:

> In the setup process you will be asked if you would like to save your [personal API Token](https://circleci.com/docs/managing-api-tokens/) into an orb-publishing [context](https://circleci.com/docs/contexts/). Saving this token is necessary for publishing development and production versions of your orb.

I [manually created a token](https://app.circleci.com/settings/user/tokens) from my (CircleCI) account. Then I created the `orb-publishing` context and added the token there. Ideally this token shouldn't be from my account but instead it should be from the `react-native-community-bot` machine user.

In any case, unfortunately my account doesn't have sufficient access to publish orbs:

```
$ circleci orb publish dist/orb.yml react-native-community/react-native@7.2.0
Once an orb is created it cannot be deleted. Orbs are semver compliant, and each published version is immutable. Publicly 
released orbs are potential dependencies for other projects.
Therefore, allowing orb deletion would make users susceptible to unexpected loss of functionality.
Error: User does not have access to publish SemVer orbs in this namespace.
```

I found out that in order to publish an `orb` the CircleCI account [must have organization level permissions](https://circleci.com/docs/orb-author-intro/#permissions-matrix):

> Orb CLI commands are scoped to different user permission levels, set by your VCS. You are the owner of your own organization. If you are authoring or publishing orbs for a namespace owned by another organization, you may require assistance from your organization admin

That means, if I can have organization level permissions I will be able to publish the 2 missing orb versions easily.

Alternatively we need a token from an "owner". I'm pretty sure that `react-native-community-orb` is listed as an organization owner for exactly this reason. I considering this the best way forward.

## Semantic Releases

Similarly for `semantic-release`, a `GH_TOKEN` is required. That personal access token (PAT) should belong to a user who has `admin` access to the repository and can create PATs scoped to the organization. Again, I'm pretty sure the `react-native-community-bot` is such a user. The bot is certainly an admin on the repository:

admin

If we login as the bot we can create a fine-grained PAT with the [necessary permissions](https://github.com/semantic-release/semantic-release/issues/2469#issuecomment-1593078854) to push tags:

* Contents: read/write
* Issues: read/write
* Metadata: read

The generated token can be stored as an environment variable as `GH_TOKEN` or added to the `orb-publishing` context (and the `node/run` job changed to use the `orb-publishing` context).

## Summary

- We need access to `react-native-community-bot` or a user who is organization owner and admin at the repository.
- From the bot (or the user) we need a CircleCI token which needs to be added to `orb-publishing` as `CIRCLE_TOKEN`. We can use that token to publish the missing versions (`v7.2.0` and `v.7.3.0`) easily using the command posted above.
- From the bot (or the user) we need a GH (fine-grained) PAT which can be used by `semantic-release`.

It's better (more secure) to use the `react-native-community-bot` user because it is (implicitly) scoped to the `react-native-community` org. If I use a personal CircleCI token (e.g. mine or yours) and we are also members of other organizations (we are), and the token leaks it can be used across orgs – that's a bad idea.

In short… I believe if I can get access to the `react-native-community-bot` account I'll be able to do all the above: publish missing orbs, fix the CI/CD for publishing the orb going forward, and configure the pipeline for doing semantic release automatically (and perhaps restore order to the galaxy).

Got any ideas how we can access the bot account? Ask Matt? Other members of the community?

**Comment by cortinico** [comment]
> Alternatively we need a token from an "owner". I'm pretty sure that `react-native-community-orb` is listed as an organization owner for exactly this reason. I considering this the best way forward.

We don't have access to it. I guess the best would be to create a new bot account (`react-native-circleci-orb-bot`) and reach out to CircleCI to move the ownership

**Comment by fotos** [comment]
> > Alternatively we need a token from an "owner". I'm pretty sure that `react-native-community-orb` is listed as an organization owner for exactly this reason. I considering this the best way forward.
> 
> We don't have access to it. I guess the best would be to create a new bot account (`react-native-circleci-orb-bot`) and reach out to CircleCI to move the ownership

Cool – no worries. I'll create a new machine user now and we can add to the org and the project. That's also a very reasonable approach if we don't have access to `react-native-community-orb`. On it.

**Comment by fotos** [comment]
@cortinico I created [`react-native-circleci-orb-bot`](https://github.com/react-native-circleci-orb-bot). I'll create the necessary GH PAT and add the bot as an admin to the project. Can you add the bot as an org owner too so it can publish orbs?

**Comment by fotos** [comment]
> and reach out to CircleCI to move the ownership

I don't believe we need to reach out to CircleCI to move the ownership. CircleCI is assigning ownership based on the GH permissions. If the new `react-native-circleci-orb-bot` is made a `react-native-community`  organization owner then it will be able to publish the orb in CircleCI. That's my current understanding and expectation.

Anyway the new bot is ready. I'll wait for the org ownership and then I'll proceed with the steps I listed above.

I'll be around all day today. 😄

**Comment by cortinico** [comment]
> Anyway the new bot is ready. I'll wait for the org ownership and then I'll proceed with the steps I listed above.

Yup I've sent an invite to the bot.
Could you also share the bot credentials with me (my contacts are available on my GH profile)?

**Comment by fotos** [comment]
> > Anyway the new bot is ready. I'll wait for the org ownership and then I'll proceed with the steps I listed above.
> 
> Yup I've sent an invite to the bot. Could you also share the bot credentials with me (my contacts are available on my GH profile)?

Accepted the invite. Sent you the credentials. ✅

**Comment by fotos** [comment]
@cortinico OK, the `CIRCLE_TOKEN` from `react-native-circleci-orb-bot` worked. I just published the pending versions and closed the other issue. The token is added in the `orb-publishing` context and the context is restricted to this repository.

Now I'll figure out the `GH_TOKEN` for `semantic-release` and might merge a (simple) PR, like the dependabot ones, to trigger the CI/CD pipeline.

**Comment by fotos** [comment]
@cortinico the `GH_TOKEN` is also in place. I haven't tried doing a release but I expect that the next `feat` or `fix` merge will automatically trigger a new release. If not let me know and I can take another look. PRs / commits need to follow conventional commits.

My part here is done for now. 😄

**Comment by cortinico** [comment]
> My part here is done for now. 😄

Amazing work @fotos Thanks for unblocking this situation!
